File: ports/zig/src/graphics/chart.zig
//! Charts of arbitrary (x, y) data.
//!
//! `plot` takes `[]const f64` and puts one sample per column: the x axis is the
//! slice index. That is the right model for a history buffer and the wrong one
//! for everything else -- two series of different lengths silently render at
//! different horizontal scales, a gap in the data is indistinguishable from a
//! shorter series, and there is no way at all to say where on the x axis a
//! point belongs.
//!
//! This takes points and a domain for each axis, so a series is placed rather
//! than appended. `plot` is untouched and still means what it meant.

const std = @import("std");

const blocks = @import("blocks.zig");
const braille_mod = @import("braille.zig");
const buffer_mod = @import("../buffer.zig");
const color_mod = @import("../color.zig");
const plot_mod = @import("plot.zig");
const surface_mod = @import("../surface.zig");
const theme_mod = @import("../theme.zig");

const BrailleCanvas = braille_mod.BrailleCanvas;
const Color = color_mod.Color;
const FillMode = blocks.FillMode;
const Point = braille_mod.Point;
const Style = buffer_mod.Style;
const Surface = surface_mod.Surface;
const clamp01 = blocks.clamp01;
const firstCodepoint = blocks.firstCodepoint;
const roundHalfUp = color_mod.roundHalfUp;
const seriesColor = theme_mod.seriesColor;

/// How a series is marked: joined, dotted, or dropped to the baseline.
pub const MarkType = enum { line, scatter, bar };

pub const ChartSeries = struct {
    points: []const Point = &.{},
    color: ?Color = null,
    label: []const u8 = "",
    mark: MarkType = .line,
    /// Shade between the line and the baseline. Ignored for a scatter.
    fill: bool = false,
};

/// One axis: what it spans and how its numbers read.
pub const AxisOptions = struct {
    min: ?f64 = null,
    max: ?f64 = null,
    format: ?*const fn (f64, []u8) []const u8 = null,
    /// How many labels to place. Default 2 -- the ends.
    ticks: usize = 0,
};

pub const ChartPlotOptions = struct {
    /// Braille is sharpest; block and ascii are the graceful degradations.
    mode: FillMode = .braille,
    x: ?AxisOptions = null,
    y: ?AxisOptions = null,
    background: ?Color = null,
    grid: bool = false,
    grid_color: ?Color = null,
    /// 0-1 opacity of the area fill against the background.
    fill_alpha: ?f64 = null,
    /// Where a bar or an area is measured from. Defaults to the y minimum.
    baseline: ?f64 = null,
};

pub const Domain = struct { min: f64, max: f64 };

/// A finite number, or null: a caller's bound is data, and data can be NaN.
fn bound(value: ?f64) ?f64 {
    const v = value orelse return null;
    return if (std.math.isFinite(v)) v else null;
}

/// The span an axis covers, from the caller where they said and from the data
/// where they did not.
///
/// A domain of zero width cannot be mapped -- every point would land in the same
/// place and a division would blow up -- so a flat series is given room around
/// itself rather than being collapsed onto one line.
pub fn domainOf(series: []const ChartSeries, axis: ?AxisOptions, which: usize) Domain {
    var min = if (axis) |a| bound(a.min) else null;
    var max = if (axis) |a| bound(a.max) else null;
    if (min == null or max == null) {
        var lo: f64 = std.math.inf(f64);
        var hi: f64 = -std.math.inf(f64);
        for (series) |s| {
            for (s.points) |p| {
                const v = if (which == 0) p.x else p.y;
                if (!std.math.isFinite(v)) continue;
                if (v < lo) lo = v;
                if (v > hi) hi = v;
            }
        }
        if (!std.math.isFinite(lo)) {
            lo = 0;
            hi = 1;
        }
        if (min == null) min = lo;
        if (max == null) max = hi;
    }
    const lo = min.?;
    const hi = max.?;
    if (!(hi > lo)) {
        // A flat series still has to be drawn somewhere sensible.
        const pad: f64 = if (@abs(lo) > 0) @abs(lo) * 0.5 else 0.5;
        return .{ .min = lo - pad, .max = lo + pad };
    }
    return .{ .min = lo, .max = hi };
}

/// Where a value sits in its domain, 0 at the minimum and 1 at the maximum.
fn ratio(value: f64, d: Domain) f64 {
    return (value - d.min) / (d.max - d.min);
}

fn drawGrid(s: Surface, color: Color, bg: ?Color) void {
    const w = s.width();
    const h = s.height();
    const step = @max(2, h / 4);
    var y: usize = 0;
    while (y < h) : (y += step) {
        var x: usize = 0;
        while (x < w) : (x += 2) {
            s.glyph(@intCast(x), @intCast(y), '·', .{ .fg = color, .bg = bg });
        }
    }
}

/// Draw point series across the whole surface.
///
/// Points are drawn in the order they are given: a line joins them as they come,
/// which is what lets a chart draw a loop or a path that doubles back. Sorting
/// them would quietly make that impossible.
pub fn plotPoints(
    allocator: std.mem.Allocator,
    s: Surface,
    series: []const ChartSeries,
    options: ChartPlotOptions,
) !void {
    if (s.isEmpty() or series.len == 0) return;
    const theme = s.theme;
    const bg = options.background;
    const w = s.width();
    const h = s.height();

    const xd = domainOf(series, options.x, 0);
    const yd = domainOf(series, options.y, 1);
    const baseline = bound(options.baseline) orelse yd.min;

    if (options.grid) {
        const color = options.grid_color orelse theme.border.mix(theme.background, 0.4);
        drawGrid(s, color, bg);
    }

    if (options.mode != .braille) {
        plotCells(s, series, options.mode, xd, yd, baseline, bg);
        return;
    }

    var canvas = try BrailleCanvas.init(allocator, w, h);
    defer canvas.deinit();
    const px: f64 = @floatFromInt(canvas.width);
    const py: f64 = @floatFromInt(canvas.height);

    for (series, 0..) |cs, si| {
        canvas.clear();
        const color = cs.color orelse seriesColor(theme.*, @intCast(si));
        if (cs.points.len == 0) continue;

        // Two buffers per series rather than one shared one: a series is drawn
        // and blitted before the next is touched, so nothing outlives the loop.
        const finite = try allocator.alloc(Point, cs.points.len);
        defer allocator.free(finite);
        var count: usize = 0;
        for (cs.points) |p| {
            if (!std.math.isFinite(p.x) or !std.math.isFinite(p.y)) continue;
            finite[count] = p;
            count += 1;
        }
        if (count == 0) continue;

        const pixels = try allocator.alloc(Point, count);
        defer allocator.free(pixels);
        for (finite[0..count], 0..) |p, i| {
            pixels[i] = .{
                .x = roundHalfUp(clamp01(ratio(p.x, xd)) * (px - 1)),
                .y = roundHalfUp((1 - clamp01(ratio(p.y, yd))) * (py - 1)),
            };
        }

        switch (cs.mark) {
            .scatter => for (pixels) |p| canvas.pixel(p.x, p.y),
            .bar => {
                const floor = roundHalfUp((1 - clamp01(ratio(baseline, yd))) * (py - 1));
                for (pixels) |p| canvas.vline(p.x, @min(p.y, floor), @max(p.y, floor));
            },
            .line => {
                if (pixels.len == 1) {
                    canvas.pixel(pixels[0].x, pixels[0].y);
                } else {
                    canvas.polyline(pixels);
                }
            },
        }

        if (cs.fill and cs.mark != .scatter) {
            try fillUnder(
                allocator,
                s,
                finite[0..count],
                xd,
                yd,
                baseline,
                color,
                bg,
                options.fill_alpha orelse 0.5,
            );
        }
        plot_mod.blitFlat(s, &canvas, color, bg);
    }
}

/// The area between a series and its baseline, in block elements.
///
/// Braille would give eight scattered dots per cell, which reads as noise where
/// an area should read as an area. The line itself stays Braille, so it keeps
/// the sub-cell resolution.
///
/// The height of each column is interpolated along the line rather than sampled
/// from the points that happen to land in it. Sampling leaves a gap wherever a
/// column has no point of its own, which with arbitrary x values is most of
/// them -- the area comes out striped instead of solid.
fn fillUnder(
    allocator: std.mem.Allocator,
    s: Surface,
    points: []const Point,
    xd: Domain,
    yd: Domain,
    baseline: f64,
    color: Color,
    bg: ?Color,
    alpha: f64,
) !void {
    const w = s.width();
    const h = s.height();
    if (w == 0 or h == 0 or points.len == 0) return;
    const base = bg orelse s.theme.background;
    const floor = clamp01(ratio(baseline, yd));

    const tops = try allocator.alloc(f64, w);
    defer allocator.free(tops);
    @memset(tops, std.math.nan(f64));

    const fw: f64 = @floatFromInt(w);
    const fh: f64 = @floatFromInt(h);

    for (0..points.len - 1) |i| {
        const p0 = points[i];
        const p1 = points[i + 1];
        const c0 = ratio(p0.x, xd) * (fw - 1);
        const c1 = ratio(p1.x, xd) * (fw - 1);
        const lo = @max(0, @floor(@min(c0, c1)));
        const hi = @min(fw - 1, @max(0, @ceil(@max(c0, c1))));
        var col: usize = @intFromFloat(lo);
        const last: usize = @intFromFloat(hi);
        while (col <= last and col < w) : (col += 1) {
            const fc: f64 = @floatFromInt(col);
            const t = if (c1 == c0) 0 else (fc - c0) / (c1 - c0);
            if (t < -0.5 or t > 1.5) continue;
            const y = p0.y + (p1.y - p0.y) * clamp01(t);
            const value = clamp01(ratio(y, yd));
            // A path that doubles back covers a column twice; the outer edge is
            // the one that bounds the area.
            const previous = tops[col];
            if (std.math.isNan(previous) or @abs(value - floor) > @abs(previous - floor)) {
                tops[col] = value;
            }
        }
    }
    if (points.len == 1) {
        const c: f64 = roundHalfUp(ratio(points[0].x, xd) * (fw - 1));
        if (c >= 0 and c < fw) tops[@intFromFloat(c)] = clamp01(ratio(points[0].y, yd));
    }

    for (0..w) |x| {
        const top = tops[x];
        if (std.math.isNan(top)) continue;
        const from01 = @min(floor, top);
        const filled = (@max(floor, top) - from01) * fh;
        const bottom: isize = @intFromFloat(@floor(from01 * fh));
        const full: isize = @intFromFloat(@floor(filled));
        var k: isize = 0;
        while (k < full and k < @as(isize, @intCast(h))) : (k += 1) {
            const row = @as(isize, @intCast(h)) - 1 - bottom - k;
            if (row < 0 or row >= @as(isize, @intCast(h))) continue;
            const depth: f64 = if (h <= 1) 0 else @as(f64, @floatFromInt(row)) / (fh - 1);
            s.glyph(@intCast(x), row, '█', .{
                .fg = base.mix(color, alpha * (1 - depth * 0.3)),
                .bg = bg,
            });
        }
        if (full < @as(isize, @intCast(h))) {
            const glyph = blocks.verticalGlyph(filled - @as(f64, @floatFromInt(full)), .block);
            const row = @as(isize, @intCast(h)) - 1 - bottom - full;
            if (!std.mem.eql(u8, glyph, " ") and row >= 0 and row < @as(isize, @intCast(h))) {
                const depth: f64 = if (h <= 1) 0 else @as(f64, @floatFromInt(row)) / (fh - 1);
                s.glyph(@intCast(x), row, firstCodepoint(glyph), .{
                    .fg = base.mix(color, alpha * (1 - depth * 0.3) + 0.12),
                    .bg = bg,
                });
            }
        }
    }
}

/// The block and ascii degradations: one column per cell, tallest point wins.
///
/// A scatter keeps its dots rather than growing columns, because a scatter that
/// fills to the baseline is a bar chart wearing the wrong name.
fn plotCells(
    s: Surface,
    series: []const ChartSeries,
    mode: FillMode,
    xd: Domain,
    yd: Domain,
    baseline: f64,
    bg: ?Color,
) void {
    const w = s.width();
    const h = s.height();
    if (w == 0 or h == 0) return;
    const theme = s.theme;
    const floor_ratio = clamp01(ratio(baseline, yd));
    const fw: f64 = @floatFromInt(w);
    const fh: f64 = @floatFromInt(h);

    // Bounded by the surface width, which is a terminal's, so the stack is the
    // right place for it and nothing here has to allocate.
    var tops: [1024]f64 = undefined;
    const columns = @min(w, tops.len);

    for (series, 0..) |cs, si| {
        const color = cs.color orelse seriesColor(theme.*, @intCast(si));
        @memset(tops[0..columns], std.math.nan(f64));

        // Highest value per column, so a column shows the peak that fell in it
        // rather than whichever point happened to be last.
        for (cs.points) |p| {
            if (!std.math.isFinite(p.x) or !std.math.isFinite(p.y)) continue;
            const raw = roundHalfUp(ratio(p.x, xd) * (fw - 1));
            const clamped = @max(0, @min(raw, fw - 1));
            const col: usize = @intFromFloat(clamped);
            if (col >= columns) continue;
            const value = clamp01(ratio(p.y, yd));
            if (std.math.isNan(tops[col]) or value > tops[col]) tops[col] = value;
        }

        for (0..columns) |x| {
            const top = tops[x];
            if (std.math.isNan(top)) continue;
            if (cs.mark == .scatter) {
                const k = @min(@as(isize, @intFromFloat(@floor(top * fh))), @as(isize, @intCast(h)) - 1);
                const row = @as(isize, @intCast(h)) - 1 - k;
                const glyph: u21 = if (mode == .ascii) '*' else '•';
                s.glyph(@intCast(x), row, glyph, .{ .fg = color, .bg = bg });
                continue;
            }
            const from = @min(floor_ratio, top) * fh;
            const filled = (@max(floor_ratio, top) - @min(floor_ratio, top)) * fh;
            const full: isize = @intFromFloat(@floor(filled));
            const bottom: isize = @intFromFloat(@floor(from));
            var k: isize = 0;
            while (k < full) : (k += 1) {
                const row = @as(isize, @intCast(h)) - 1 - bottom - k;
                if (row >= 0 and row < @as(isize, @intCast(h))) {
                    s.glyph(@intCast(x), row, '█', .{ .fg = color, .bg = bg });
                }
            }
            const glyph = blocks.verticalGlyph(filled - @as(f64, @floatFromInt(full)), mode);
            const row = @as(isize, @intCast(h)) - 1 - bottom - full;
            if (!std.mem.eql(u8, glyph, " ") and row >= 0 and row < @as(isize, @intCast(h))) {
                s.glyph(@intCast(x), row, firstCodepoint(glyph), .{ .fg = color, .bg = bg });
            }
        }
    }
}

File: ports/rust/src/graphics/chart.rs
//! Charts of arbitrary (x, y) data.
//!
//! `plot` takes `&[f64]` and puts one sample per column: the x axis is the
//! array index. That is the right model for a history buffer and the wrong one
//! for everything else -- two series of different lengths silently render at
//! different horizontal scales, a gap in the data is indistinguishable from a
//! shorter series, and there is no way at all to say where on the x axis a
//! point belongs.
//!
//! This takes points and a domain for each axis, so a series is placed rather
//! than appended. `plot` is untouched and still means what it meant.

use crate::buffer::Style;
use crate::color::{round_half_up, Color};
use crate::graphics::blocks::{vertical_glyph, FillMode};
use crate::graphics::braille::BrailleCanvas;
use crate::graphics::plot::{blit, first_char};
use crate::surface::Surface;
use crate::theme::series_color;

pub type Point = (f64, f64);

/// How a series is marked: joined, dotted, or dropped to the baseline.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum MarkType {
    #[default]
    Line,
    Scatter,
    Bar,
}

#[derive(Clone, Debug, Default)]
pub struct ChartSeries {
    pub points: Vec<Point>,
    pub color: Option<Color>,
    pub label: Option<String>,
    pub mark: MarkType,
    /// Shade between the line and the baseline. Ignored for a scatter.
    pub fill: bool,
}

impl ChartSeries {
    pub fn new(points: impl Into<Vec<Point>>) -> ChartSeries {
        ChartSeries { points: points.into(), ..Default::default() }
    }

    pub fn mark(mut self, mark: MarkType) -> ChartSeries {
        self.mark = mark;
        self
    }

    pub fn label(mut self, label: impl Into<String>) -> ChartSeries {
        self.label = Some(label.into());
        self
    }

    pub fn filled(mut self) -> ChartSeries {
        self.fill = true;
        self
    }
}

/// One axis: what it spans and how its numbers read.
#[derive(Clone, Default)]
pub struct AxisOptions {
    pub min: Option<f64>,
    pub max: Option<f64>,
    pub format: Option<std::sync::Arc<dyn Fn(f64) -> String + Send + Sync>>,
    /// How many labels to place. Default 2 -- the ends.
    pub ticks: Option<usize>,
}

impl std::fmt::Debug for AxisOptions {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("AxisOptions")
            .field("min", &self.min)
            .field("max", &self.max)
            .field("ticks", &self.ticks)
            .finish()
    }
}

#[derive(Clone, Debug, Default)]
pub struct ChartPlotOptions {
    /// Braille is sharpest; block and ascii are the graceful degradations.
    pub mode: Option<FillMode>,
    pub x: Option<AxisOptions>,
    pub y: Option<AxisOptions>,
    pub background: Option<Color>,
    pub grid: bool,
    pub grid_color: Option<Color>,
    /// 0-1 opacity of the area fill against the background.
    pub fill_alpha: Option<f64>,
    /// Where a bar or an area is measured from. Defaults to the y minimum.
    pub baseline: Option<f64>,
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Domain {
    pub min: f64,
    pub max: f64,
}

/// A finite number, or none: a caller's bound is data, and data can be NaN.
fn bound(value: Option<f64>) -> Option<f64> {
    value.filter(|v| v.is_finite())
}

/// The span an axis covers, from the caller where they said and from the data
/// where they did not.
///
/// A domain of zero width cannot be mapped -- every point would land in the
/// same place and a division would blow up -- so a flat series is given room
/// around itself rather than being collapsed onto one line.
pub fn domain_of(series: &[ChartSeries], axis: Option<&AxisOptions>, which: usize) -> Domain {
    let mut min = bound(axis.and_then(|a| a.min));
    let mut max = bound(axis.and_then(|a| a.max));
    if min.is_none() || max.is_none() {
        let mut lo = f64::INFINITY;
        let mut hi = f64::NEG_INFINITY;
        for s in series {
            for p in &s.points {
                let v = if which == 0 { p.0 } else { p.1 };
                if !v.is_finite() {
                    continue;
                }
                if v < lo {
                    lo = v;
                }
                if v > hi {
                    hi = v;
                }
            }
        }
        if !lo.is_finite() {
            lo = 0.0;
            hi = 1.0;
        }
        min = min.or(Some(lo));
        max = max.or(Some(hi));
    }
    let min = min.unwrap_or(0.0);
    let max = max.unwrap_or(1.0);
    if !(max > min) {
        // A flat series still has to be drawn somewhere sensible.
        let pad = if min.abs() > 0.0 { min.abs() * 0.5 } else { 0.5 };
        return Domain { min: min - pad, max: min + pad };
    }
    Domain { min, max }
}

/// Where a value sits in its domain, 0 at the minimum and 1 at the maximum.
fn ratio(value: f64, domain: Domain) -> f64 {
    (value - domain.min) / (domain.max - domain.min)
}

fn clamp01(v: f64) -> f64 {
    v.clamp(0.0, 1.0)
}

fn draw_grid(surface: &Surface, color: Color, bg: Option<Color>) {
    let w = surface.width();
    let h = surface.height();
    let step = std::cmp::max(2, h / 4);
    let mut y = 0;
    while y < h {
        let mut x = 0;
        while x < w {
            surface.glyph(x as isize, y as isize, '·', &Style { fg: Some(color), bg, attrs: None });
            x += 2;
        }
        y += step;
    }
}

/// Draw point series across the whole surface.
///
/// Points are drawn in the order they are given: a line joins them as they
/// come, which is what lets a chart draw a loop or a path that doubles back.
/// Sorting them would quietly make that impossible.
pub fn plot_points(surface: &Surface, series: &[ChartSeries], options: &ChartPlotOptions) {
    if surface.is_empty() || series.is_empty() {
        return;
    }
    let theme = surface.theme.clone();
    let mode = options.mode.unwrap_or(FillMode::Braille);
    let bg = options.background;
    let w = surface.width();
    let h = surface.height();

    let xd = domain_of(series, options.x.as_ref(), 0);
    let yd = domain_of(series, options.y.as_ref(), 1);
    let baseline = bound(options.baseline).unwrap_or(yd.min);

    if options.grid {
        let color = options
            .grid_color
            .unwrap_or_else(|| theme.border.mix(theme.background, 0.4));
        draw_grid(surface, color, bg);
    }

    if mode != FillMode::Braille {
        plot_cells(surface, series, mode, xd, yd, baseline, bg);
        return;
    }

    let mut canvas = BrailleCanvas::new(w, h);
    let px = canvas.width as f64;
    let py = canvas.height as f64;

    for (si, s) in series.iter().enumerate() {
        canvas.clear();
        let color = s.color.unwrap_or_else(|| series_color(&theme, si as i64));
        let finite: Vec<Point> = s
            .points
            .iter()
            .copied()
            .filter(|p| p.0.is_finite() && p.1.is_finite())
            .collect();
        if finite.is_empty() {
            continue;
        }
        let pixels: Vec<(f64, f64)> = finite
            .iter()
            .map(|p| {
                (
                    round_half_up(clamp01(ratio(p.0, xd)) * (px - 1.0)),
                    round_half_up((1.0 - clamp01(ratio(p.1, yd))) * (py - 1.0)),
                )
            })
            .collect();

        match s.mark {
            MarkType::Scatter => {
                for (x, y) in &pixels {
                    canvas.pixel(*x, *y);
                }
            }
            MarkType::Bar => {
                let floor = round_half_up((1.0 - clamp01(ratio(baseline, yd))) * (py - 1.0));
                for (x, y) in &pixels {
                    canvas.vline(*x, y.min(floor), y.max(floor));
                }
            }
            MarkType::Line => {
                if pixels.len() == 1 {
                    canvas.pixel(pixels[0].0, pixels[0].1);
                } else {
                    canvas.polyline(&pixels);
                }
            }
        }

        if s.fill && s.mark != MarkType::Scatter {
            fill_under(
                surface,
                &finite,
                xd,
                yd,
                baseline,
                color,
                bg,
                options.fill_alpha.unwrap_or(0.5),
            );
        }
        blit(surface, &canvas, |_, _| color, bg);
    }
}

/// The area between a series and its baseline, in block elements.
///
/// Braille would give eight scattered dots per cell, which reads as noise where
/// an area should read as an area. The line itself stays Braille, so it keeps
/// the sub-cell resolution.
///
/// The height of each column is interpolated along the line rather than sampled
/// from the points that happen to land in it. Sampling leaves a gap wherever a
/// column has no point of its own, which with arbitrary x values is most of
/// them -- the area comes out striped instead of solid.
#[allow(clippy::too_many_arguments)]
fn fill_under(
    surface: &Surface,
    points: &[Point],
    xd: Domain,
    yd: Domain,
    baseline: f64,
    color: Color,
    bg: Option<Color>,
    alpha: f64,
) {
    let w = surface.width();
    let h = surface.height();
    if w == 0 || h == 0 || points.is_empty() {
        return;
    }
    let base = bg.unwrap_or(surface.theme.background);
    let floor = clamp01(ratio(baseline, yd));
    let column = |x: f64| ratio(x, xd) * (w as f64 - 1.0);

    let mut tops = vec![f64::NAN; w];
    let mut record = |col: isize, value: f64, tops: &mut Vec<f64>| {
        if col < 0 || col as usize >= tops.len() {
            return;
        }
        // A path that doubles back covers a column twice; the outer edge is the
        // one that bounds the area.
        let previous = tops[col as usize];
        if previous.is_nan() || (value - floor).abs() > (previous - floor).abs() {
            tops[col as usize] = value;
        }
    };

    if points.len() == 1 {
        record(round_half_up(column(points[0].0)) as isize, clamp01(ratio(points[0].1, yd)), &mut tops);
    }
    for pair in points.windows(2) {
        let (x0, y0) = pair[0];
        let (x1, y1) = pair[1];
        let c0 = column(x0);
        let c1 = column(x1);
        let from = c0.min(c1).floor().max(0.0) as usize;
        let to = std::cmp::min(w - 1, c0.max(c1).ceil().max(0.0) as usize);
        for col in from..=to {
            let t = if c1 == c0 { 0.0 } else { (col as f64 - c0) / (c1 - c0) };
            if !(-0.5..=1.5).contains(&t) {
                continue;
            }
            let y = y0 + (y1 - y0) * clamp01(t);
            record(col as isize, clamp01(ratio(y, yd)), &mut tops);
        }
    }

    for x in 0..w {
        let top = tops[x];
        if top.is_nan() {
            continue;
        }
        let from01 = floor.min(top);
        let filled = (floor.max(top) - from01) * h as f64;
        let bottom = (from01 * h as f64).floor() as isize;
        let full = filled.floor() as isize;
        for k in 0..full.min(h as isize) {
            let row = h as isize - 1 - bottom - k;
            if row < 0 || row >= h as isize {
                continue;
            }
            let depth = if h <= 1 { 0.0 } else { row as f64 / (h as f64 - 1.0) };
            surface.glyph(
                x as isize,
                row,
                '█',
                &Style {
                    fg: Some(base.mix(color, alpha * (1.0 - depth * 0.3))),
                    bg,
                    attrs: None,
                },
            );
        }
        if full < h as isize {
            let glyph = vertical_glyph(filled - full as f64, FillMode::Block);
            let row = h as isize - 1 - bottom - full;
            if glyph != " " && row >= 0 && row < h as isize {
                let depth = if h <= 1 { 0.0 } else { row as f64 / (h as f64 - 1.0) };
                surface.glyph(
                    x as isize,
                    row,
                    first_char(glyph),
                    &Style {
                        fg: Some(base.mix(color, alpha * (1.0 - depth * 0.3) + 0.12)),
                        bg,
                        attrs: None,
                    },
                );
            }
        }
    }
}

/// The block and ascii degradations: one column per cell, tallest point wins.
///
/// A scatter keeps its dots rather than growing columns, because a scatter that
/// fills to the baseline is a bar chart wearing the wrong name.
fn plot_cells(
    surface: &Surface,
    series: &[ChartSeries],
    mode: FillMode,
    xd: Domain,
    yd: Domain,
    baseline: f64,
    bg: Option<Color>,
) {
    let w = surface.width();
    let h = surface.height();
    let theme = surface.theme.clone();
    let floor_ratio = clamp01(ratio(baseline, yd));

    for (si, s) in series.iter().enumerate() {
        let color = s.color.unwrap_or_else(|| series_color(&theme, si as i64));
        // Highest value per column, so a column shows the peak that fell in it
        // rather than whichever point happened to be last.
        let mut tops = vec![f64::NAN; w];
        for p in &s.points {
            if !p.0.is_finite() || !p.1.is_finite() {
                continue;
            }
            let col = (round_half_up(ratio(p.0, xd) * (w as f64 - 1.0)) as isize)
                .clamp(0, w as isize - 1) as usize;
            let value = clamp01(ratio(p.1, yd));
            if tops[col].is_nan() || value > tops[col] {
                tops[col] = value;
            }
        }

        for x in 0..w {
            let top = tops[x];
            if top.is_nan() {
                continue;
            }
            if s.mark == MarkType::Scatter {
                let row = h as isize - 1 - ((top * h as f64).floor() as isize).min(h as isize - 1);
                let glyph = if mode == FillMode::Ascii { '*' } else { '•' };
                surface.glyph(x as isize, row, glyph, &Style { fg: Some(color), bg, attrs: None });
                continue;
            }
            let from = floor_ratio.min(top) * h as f64;
            let filled = (floor_ratio.max(top) - floor_ratio.min(top)) * h as f64;
            let full = filled.floor() as isize;
            for k in 0..full {
                let row = h as isize - 1 - from.floor() as isize - k;
                if row >= 0 && row < h as isize {
                    surface.glyph(x as isize, row, '█', &Style { fg: Some(color), bg, attrs: None });
                }
            }
            let glyph = vertical_glyph(filled - full as f64, mode);
            let row = h as isize - 1 - from.floor() as isize - full;
            if glyph != " " && row >= 0 && row < h as isize {
                surface.glyph(x as isize, row, first_char(glyph), &Style { fg: Some(color), bg, attrs: None });
            }
        }
    }
}
